### ISSUE TITLE
Slack.Engine could not start

### ISSUE THREAD

**Issue by Auha** [saltstack/salt#46943]
### Description of Issue/Question

I recently upgraded to Salt version 2018.3.0 on my master, and then I decided to try to utilize salt engine. I would get the following errors in my log immediately whenI try to do a simple command from a salt channel.

```
2018-04-05 16:58:11,467 [tornado.application:612 ][ERROR   ][23660] Exception in callback <functools.partial object at 0x51ff310>
Traceback (most recent call last):
  File "/usr/lib64/python2.7/site-packages/tornado/ioloop.py", line 592, in _run_callback
    ret = callback()
  File "/usr/lib64/python2.7/site-packages/tornado/stack_context.py", line 275, in null_wrapper
    return fn(*args, **kwargs)
  File "/usr/lib64/python2.7/site-packages/tornado/ioloop.py", line 598, in <lambda>
    self.add_future(ret, lambda f: f.result())
  File "/usr/lib64/python2.7/site-packages/tornado/concurrent.py", line 215, in result
    raise_exc_info(self._exc_info)
  File "/usr/lib64/python2.7/site-packages/tornado/gen.py", line 230, in wrapper
    yielded = next(result)
  File "/usr/lib/python2.7/site-packages/salt/crypt.py", line 596, in _authenticate
    io_loop=self.io_loop)
  File "/usr/lib/python2.7/site-packages/salt/transport/client.py", line 108, in factory
    return salt.transport.zeromq.AsyncZeroMQReqChannel(opts, **kwargs)
  File "/usr/lib/python2.7/site-packages/salt/transport/zeromq.py", line 132, in __new__
    obj.__singleton_init__(opts, **kwargs)
  File "/usr/lib/python2.7/site-packages/salt/transport/zeromq.py", line 196, in __singleton_init__
    kwargs={'io_loop': self._io_loop})
  File "/usr/lib/python2.7/site-packages/salt/transport/zeromq.py", line 929, in __init__
    super(AsyncReqMessageClientPool, self).__init__(AsyncReqMessageClient, opts, args=args, kwargs=kwargs)
  File "/usr/lib/python2.7/site-packages/salt/transport/__init__.py", line 71, in __init__
    self.message_clients = [tgt(*args, **kwargs) for _ in range(sock_pool_size)]
  File "/usr/lib/python2.7/site-packages/salt/transport/zeromq.py", line 976, in __init__
    self._init_socket()
  File "/usr/lib/python2.7/site-packages/salt/transport/zeromq.py", line 1030, in _init_socket
    self.stream = zmq.eventloop.zmqstream.ZMQStream(self.socket, io_loop=self.io_loop)
  File "/usr/lib64/python2.7/site-packages/zmq/eventloop/zmqstream.py", line 114, in __init__
    self._init_io_state()
  File "/usr/lib64/python2.7/site-packages/zmq/eventloop/zmqstream.py", line 535, in _init_io_state
    self.io_loop.add_handler(self.socket, self._handle_events, self._state)
  File "/usr/lib64/python2.7/site-packages/tornado/ioloop.py", line 704, in add_handler
    self._impl.register(fd, events | self.ERROR)
TypeError: argument must be an int, or have a fileno() method.
```
Afterwards, I would work on something else and come back a few hours later, and I would get this in the logs.
```
2018-04-05 17:03:16,937 [salt.engines     :132 ][CRITICAL][24907] Engine 'slack' could not be started!
Traceback (most recent call last):
  File "/usr/lib/python2.7/site-packages/salt/engines/__init__.py", line 128, in run
    self.engine[self.fun](**kwargs)
  File "/usr/lib/python2.7/site-packages/salt/engines/slack.py", line 777, in start
    raise Exception('{}'.format(traceback.format_exc()))
Exception: Traceback (most recent call last):
  File "/usr/lib/python2.7/site-packages/salt/engines/slack.py", line 775, in start
    client.run_commands_from_slack_async(message_generator, fire_all, tag, control)
  File "/usr/lib/python2.7/site-packages/salt/engines/slack.py", line 652, in run_commands_from_slack_async
    for msg in message_generator:
  File "/usr/lib/python2.7/site-packages/salt/engines/slack.py", line 462, in generate_triggered_messages
    data = just_data(m_data)
  File "/usr/lib/python2.7/site-packages/salt/engines/slack.py", line 427, in just_data
    'user_id': user_id,
UnboundLocalError: local variable 'user_id' referenced before assignment
```

### Setup
(Please provide relevant configs and/or SLS files (Be sure to remove sensitive info).)

My master config with the engines section looks like the following with the token redacted for obvious reasons:
```
engines:
  - slack:
      token: 'xoxb-xxxxx'
      control: True
      fire_all: False

      groups:
        default:
          users:
            - '*'
        commands:
          - test.ping

        default_target:
          target: salt
          tgt_type: glob

        targets:
          test.ping:
            target: '*'
            tgt_type: glob
```

### Steps to Reproduce Issue
(Include debug logs if possible and relevant.)

Look above in the description.

### Versions Report
(Provided by running `salt --versions-report`. Please also mention any differences in master/minion versions.)
```
Salt Version:
           Salt: 2018.3.0
 
Dependency Versions:
           cffi: 1.11.2
       cherrypy: Not Installed
       dateutil: 1.5
      docker-py: Not Installed
          gitdb: 0.6.4
      gitpython: 1.0.1
          ioflo: Not Installed
         Jinja2: 2.7.2
        libgit2: Not Installed
        libnacl: Not Installed
       M2Crypto: Not Installed
           Mako: Not Installed
   msgpack-pure: Not Installed
 msgpack-python: 0.5.1
   mysql-python: Not Installed
      pycparser: 2.18
       pycrypto: 2.6.1
   pycryptodome: Not Installed
         pygit2: Not Installed
         Python: 2.7.5 (default, Sep  5 2016, 02:30:38)
   python-gnupg: Not Installed
         PyYAML: 3.11
          PyZMQ: 15.3.0
           RAET: Not Installed
          smmap: 0.9.0
        timelib: Not Installed
        Tornado: 4.2.1
            ZMQ: 4.1.4
 
System Versions:
           dist: oracle 7.3 
         locale: UTF-8
        machine: x86_64
        release: 4.1.12-61.1.18.el7uek.x86_64
         system: Linux
        version: Oracle Linux Server 7.3 
```

**Comment by viq** [comment]
Looks like I'm experiencing the same `UnboundLocalError: local variable 'user_id' referenced before assignment`, though I see it immediately after starting master.
```
Salt Version:
           Salt: 2018.3.0
 
Dependency Versions:
           cffi: 1.11.5
       cherrypy: Not Installed
       dateutil: Not Installed
      docker-py: Not Installed
          gitdb: Not Installed
      gitpython: Not Installed
          ioflo: Not Installed
         Jinja2: 2.10
        libgit2: 0.26.3
        libnacl: Not Installed
       M2Crypto: Not Installed
           Mako: Not Installed
   msgpack-pure: Not Installed
 msgpack-python: 0.5.6
   mysql-python: Not Installed
      pycparser: 2.18
       pycrypto: 2.6.1
   pycryptodome: Not Installed
         pygit2: 0.26.4
         Python: 3.6.4 (default, Jan  5 2018, 02:35:40)
   python-gnupg: 0.4.2
         PyYAML: 3.12
          PyZMQ: 17.0.0
           RAET: Not Installed
          smmap: Not Installed
        timelib: Not Installed
        Tornado: 4.5.3
            ZMQ: 4.1.6
 
System Versions:
           dist: arch  
         locale: UTF-8
        machine: x86_64
        release: 4.15.15-1-hardened
         system: Linux
        version: arch  
```

**Comment by Ch3LL** [comment]
ping @garethgreenaway any ideas here?

**Comment by garethgreenaway** [comment]
@Ch3LL Looking into it

**Comment by garethgreenaway** [comment]
@Auha I think this might be the same issue as reported in #46947, the indention is off so `commands`, `aliases`, `default_target`, `targets` are not under the `default` group in the documentation and they need to be.

**Comment by Auha** [comment]
I'll try that out later tonight.

**Comment by kyrias** [comment]
I just hit the same issue trying out the Salt engine, and I have the `commands` &co indented under the group itself properly.

**Comment by garethgreenaway** [comment]
@kyrias Can you provide a scrubbed copy of your Slack engine configuration?

**Comment by kyrias** [comment]
@garethgreenaway Sure, here:

```
engines:
  - slack:
      token: '<token>'
      control: True
      fire_all: False
      groups:
        default:
          users:
            - '*'
          commands:
            - test.ping
          default_target:
            target: 'salt*'
            tgt_type: glob
```

**Comment by kyrias** [comment]
Not sure if it'll help, but the following is the debug logs from when running a slack command:

```
[DEBUG   ] Got an error from trying to get the message text This is not a message
[INFO    ] Message is !test.ping
[DEBUG   ] Requesting URL https://slack.com/api/channels.list?token=<token> using GET method
[DEBUG   ] Using backend: tornado
[DEBUG   ] Response Status Code: 200
[DEBUG   ] use_groups {u'default': {u'commands': [u'test.ping'], u'default_target': {u'tgt_type': u'glob', u'target': u'salt*'}, u'users': [u'kyrias']}}
[DEBUG   ] Missing configuration file: /etc/salt/minion
[DEBUG   ] Including configuration from '/etc/salt/minion.d/_schedule.conf'
[DEBUG   ] Reading configuration from /etc/salt/minion.d/_schedule.conf
[DEBUG   ] Including configuration from '/etc/salt/minion.d/beacons.conf'
[DEBUG   ] Reading configuration from /etc/salt/minion.d/beacons.conf
[DEBUG   ] Including configuration from '/etc/salt/minion.d/engine.conf'
[DEBUG   ] Reading configuration from /etc/salt/minion.d/engine.conf
[DEBUG   ] Including configuration from '/etc/salt/minion.d/f_defaults.conf'
[DEBUG   ] Reading configuration from /etc/salt/minion.d/f_defaults.conf
[DEBUG   ] Including configuration from '/etc/salt/minion.d/reactor.conf'
[DEBUG   ] Reading configuration from /etc/salt/minion.d/reactor.conf
[DEBUG   ] Using cached minion ID from /etc/salt/minion_id: salt.internal.example.com
[DEBUG   ] Grains refresh requested. Refreshing grains.
[DEBUG   ] Missing configuration file: /etc/salt/minion
[DEBUG   ] Including configuration from '/etc/salt/minion.d/_schedule.conf'
[DEBUG   ] Reading configuration from /etc/salt/minion.d/_schedule.conf
[DEBUG   ] Including configuration from '/etc/salt/minion.d/beacons.conf'
[DEBUG   ] Reading configuration from /etc/salt/minion.d/beacons.conf
[DEBUG   ] Including configuration from '/etc/salt/minion.d/engine.conf'
[DEBUG   ] Reading configuration from /etc/salt/minion.d/engine.conf
[DEBUG   ] Including configuration from '/etc/salt/minion.d/f_defaults.conf'
[DEBUG   ] Reading configuration from /etc/salt/minion.d/f_defaults.conf
[DEBUG   ] Including configuration from '/etc/salt/minion.d/reactor.conf'
[DEBUG   ] Reading configuration from /etc/salt/minion.d/reactor.conf
[DEBUG   ] Connecting to master. Attempt 1 of 1
[DEBUG   ] Master URI: tcp://172.19.30.10:4506
[DEBUG   ] Initializing new AsyncAuth for (u'/etc/salt/pki/minion', u'salt.internal.example.com', u'tcp://172.19.30.10:4506')
[DEBUG   ] Generated random reconnect delay between '1000ms' and '11000ms' (1894)
[DEBUG   ] Setting zmq_reconnect_ivl to '1894ms'
[DEBUG   ] Setting zmq_reconnect_ivl_max to '11000ms'
[DEBUG   ] Initializing new AsyncZeroMQReqChannel for (u'/etc/salt/pki/minion', u'salt.internal.example.com', u'tcp://172.19.30.10:4506', 'clear')
[DEBUG   ] Connecting the Minion to the Master URI (for the return server): tcp://172.19.30.10:4506
[DEBUG   ] Trying to connect to: tcp://172.19.30.10:4506
[ERROR   ] Exception in callback <functools.partial object at 0x41fa470>
Traceback (most recent call last):
  File "/usr/lib64/python2.7/site-packages/tornado/ioloop.py", line 591, in _run_callback
    ret = callback()
  File "/usr/lib64/python2.7/site-packages/tornado/stack_context.py", line 274, in null_wrapper
    return fn(*args, **kwargs)
  File "/usr/lib64/python2.7/site-packages/tornado/ioloop.py", line 597, in <lambda>
    self.add_future(ret, lambda f: f.result())
  File "/usr/lib64/python2.7/site-packages/tornado/concurrent.py", line 214, in result
    raise_exc_info(self._exc_info)
  File "/usr/lib64/python2.7/site-packages/tornado/gen.py", line 230, in wrapper
    yielded = next(result)
  File "/usr/lib/python2.7/site-packages/salt/crypt.py", line 596, in _authenticate
    io_loop=self.io_loop)
  File "/usr/lib/python2.7/site-packages/salt/transport/client.py", line 108, in factory
    return salt.transport.zeromq.AsyncZeroMQReqChannel(opts, **kwargs)
  File "/usr/lib/python2.7/site-packages/salt/transport/zeromq.py", line 132, in __new__
    obj.__singleton_init__(opts, **kwargs)
  File "/usr/lib/python2.7/site-packages/salt/transport/zeromq.py", line 196, in __singleton_init__
    kwargs={'io_loop': self._io_loop})
  File "/usr/lib/python2.7/site-packages/salt/transport/zeromq.py", line 929, in __init__
    super(AsyncReqMessageClientPool, self).__init__(AsyncReqMessageClient, opts, args=args, kwargs=kwargs)
  File "/usr/lib/python2.7/site-packages/salt/transport/__init__.py", line 71, in __init__
    self.message_clients = [tgt(*args, **kwargs) for _ in range(sock_pool_size)]
  File "/usr/lib/python2.7/site-packages/salt/transport/zeromq.py", line 976, in __init__
    self._init_socket()
  File "/usr/lib/python2.7/site-packages/salt/transport/zeromq.py", line 1030, in _init_socket
    self.stream = zmq.eventloop.zmqstream.ZMQStream(self.socket, io_loop=self.io_loop)
  File "/usr/lib64/python2.7/site-packages/zmq/eventloop/zmqstream.py", line 114, in __init__
    self._init_io_state()
  File "/usr/lib64/python2.7/site-packages/zmq/eventloop/zmqstream.py", line 535, in _init_io_state
    self.io_loop.add_handler(self.socket, self._handle_events, self._state)
  File "/usr/lib64/python2.7/site-packages/tornado/ioloop.py", line 703, in add_handler
    self._impl.register(fd, events | self.ERROR)
TypeError: argument must be an int, or have a fileno() method.
```

**Comment by sootythebear** [comment]
The `UnboundLocalError: local variable 'user_id' referenced before assignment` error is caused by a RTM message that does not include a user reference within the payload. For example, a `file_comment` message:

```
{
    "type": "message",
    "subtype": "file_comment",
    "ts": "1361482916.000003",
    "text": "<@cal> commented on a file: ...",
    "file": {},
    "comment": {}
}
```
A `bot_message` RTM message looks like a likely cause too. 

I have reproduced the Slack error with a `file_comment`.

The following is thrown just prior to Slack failing:

```
2018-04-12 15:04:16,518 [salt.loaded.int.engines.slack:370 ][INFO    ][4829] Message is <@Wxxxxxx> commented on <@Wxxxxxx>’s file <https://xxxxxx.enterprise.slack.com/files/W8HDCC93R/F8ZFFE5HT/salt-results-20180126181550678856.yaml|salt-results-20180126181550678856>: add comment
```

**Comment by alexfrosa** [comment]
Having same issue with Slack Engine, exactly salt version.

**Comment by sootythebear** [comment]
The second (`UnboundLocalError: local variable 'user_id' referenced before assignment`) is a message from Slack to Salt via RTM. See `https://api.slack.com/events/message/file_comment`, where the user field is missing. 

There is no mechanism to handle a missing `user` field within the RTM message within slack.py at lines 413-420:
```
            if 'user' not in m_data:
                if 'message' in m_data and 'user' in m_data['message']:
                    log.debug('Message was edited, '
                              'so we look for user in '
                              'the original message.')
                    user_id = m_data['message']['user']
            else:
                user_id = m_data.get('user')
```

I believe there needs to be an `else:` for the `message if` statement to capture this scenario.

**Comment by kyrias** [comment]
So this is weird, I spun up a new CentOS 7 instance and installed the master, and the slack engine *did* work there.  So I compared the versions reported by `-V`, and figured I'd go through them one-by-one until I got it working.  I removed `python-cffi` first, which also removed `python-pygit2`, and after that the IOLoop error disappeared, but installing `python-pygit2` again made it come back.

**Comment by garethgreenaway** [comment]
@sootythebear Thanks for the confirmation.  That was my conclusion as well. Working on a fix now.  Thanks! 😊

**Comment by kyrias** [comment]
(I should note that it still gives me the following error, which seems to happen when `groups_pillar_name` isn't specified.  Would be nice if it wasn't necessary to specify that key.

```
[CRITICAL] Engine 'slack' could not be started!
Traceback (most recent call last):
  File "/usr/lib/python2.7/site-packages/salt/engines/__init__.py", line 128, in run
    self.engine[self.fun](**kwargs)
  File "/usr/lib/python2.7/site-packages/salt/engines/slack.py", line 777, in start
    raise Exception('{}'.format(traceback.format_exc()))
Exception: Traceback (most recent call last):
  File "/usr/lib/python2.7/site-packages/salt/engines/slack.py", line 775, in start
    client.run_commands_from_slack_async(message_generator, fire_all, tag, control)
  File "/usr/lib/python2.7/site-packages/salt/engines/slack.py", line 652, in run_commands_from_slack_async
    for msg in message_generator:
  File "/usr/lib/python2.7/site-packages/salt/engines/slack.py", line 464, in generate_triggered_messages
    loaded_groups = self.get_config_groups(groups, groups_pillar_name)
  File "/usr/lib/python2.7/site-packages/salt/engines/slack.py", line 207, in get_config_groups
    log.warn('Failed to get groups from %s: %s', groups_pillar_name, self._groups_from_pillar(groups_pillar_name))
  File "/usr/lib/python2.7/site-packages/salt/engines/slack.py", line 239, in _groups_from_pillar
    pillar_groups = caller.cmd('pillar.get', pillar_name)
  File "/usr/lib/python2.7/site-packages/salt/client/__init__.py", line 2094, in cmd
    return self.sminion.functions[fun](*args, **kwargs)
  File "/usr/lib/python2.7/site-packages/salt/modules/pillar.py", line 175, in get
    delimiter)
  File "/usr/lib/python2.7/site-packages/salt/utils/data.py", line 463, in traverse_dict_and_list
    for each in key.split(delimiter):
AttributeError: 'NoneType' object has no attribute 'split'
```
)

**Comment by alexfrosa** [comment]
This fix worked for my Salt Master. Thank you very much!

**Comment by kyrias** [comment]
It was.

**Comment by rallytime** [comment]
Closed via #47262
